File: work/dp_e.py
# -*- coding: utf-8 -*-
N,W = map(int, input().split())
wl = []
vl = []
for _ in range(N):
    w,v = map(int, input().split())
    wl.append(w)
    vl.append(v)

INF = 10**11
V = sum(vl)+1
dp = [[INF for _ in range(V+1)] for _ in range(N+1)]
dp[0][0] = 0

for vi in range(V+1):
    for i in range(N):
        if vi>=vl[i]:
            dp[i+1][vi] = min(
                dp[i][vi],
                dp[i][vi-vl[i]]+wl[i]
            )
        else:
            dp[i+1][vi] = dp[i][vi]

res = 0
for vi in range(V+1):
    if dp[N][vi]<=W:
        res = max(res, vi)

print(res)

File: work/dp_e.cpp
#include<iostream>
#include<iomanip>
#include<vector>
#include<algorithm>
#include<string>
using namespace std;

typedef long long ll;

int main(){
  ll n,W;
  cin>>n>>W;
  vector<ll> wl = vector<ll>(n);
  vector<ll> vl = vector<ll>(n);

  for(int i=0;i<n;i++){
    ll w,v;
    cin>>w>>v;
    wl[i] = w;
    vl[i] = v;
  }

  ll INF = 1000000000*n+10;
  ll V = 1000*n+10;
  vector<vector<ll>> dp = vector<vector<ll>>(n+1, vector<ll>(V+1, INF));
  dp[0][0] = 0;
  for(int vi=0;vi<V+1;vi++){
    for(int i=0;i<n;i++){
      if(vl[i]>vi){
        dp[i+1][vi] = dp[i][vi];
      }else{
        dp[i+1][vi] = min(dp[i][vi], dp[i][vi-vl[i]]+wl[i]);
      }
    }
  }

  ll res = 0;
  for(ll vi=0;vi<V+1;vi++){
    if(dp[n][vi]<=W){
      res = max(res, vi);
    }
  }
  cout<<res<<endl;
  return 0;
}
